### DISCUSSION TITLE
Custom Repository Type (for Console)

### DISCUSSION THREAD

**Discussion by triggan** [eclipse-rdf4j/rdf4j#3795] [General]
Hi,

I'm currently working with an RDF store that uses an Authorization Header (similar to JWT or OAuth) for security.  I'm looking to architect a means by which a user can leverage the RDF4j Console to access this store.  After initial investigation, I think it would make the most sense to extend the SPARQLRepository class/type and create a new custom repository type.  This way users can use a similar SPARQLRepository template structure when defining access to the RDF store.  The new class would include parameters for API keys/tokens.

Example (note the repositoryType and added custom fields):
```
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.
@prefix rep: <http://www.openrdf.org/config/repository#>.
@prefix sparql: <http://www.openrdf.org/config/repository/sparql#>.
@prefix custom: <http://www.openrdf.org/config/repository/sparql/custom#>.

[] a rep:Repository ;
   rep:repositoryID "example" ;
   rdfs:label "SPARQL endpoint at http://example.org/" .
   rep:repositoryImpl [
      rep:repositoryType "custom:MyCustomSPARQLRepository" ;
      sparql:query-endpoint <http://example.org/sparql> ;
      sparql:update-endpoint <http://example.org/sparql/update> ;
      custom:key "abcdefghikjlmn" ;
      custom:token "12345abcde78906" ;
   ];
```

Looking for suggestions on how I would go about adding this functionality.  Or are there extensions that already exist to allow for this?

**Comment by abrokenjester** [comment]
I think that makes a lot of sense. Fwiw if you have the bandwidth to actually contribute this feature back to the open-source project, it might even be something you could just add into the existing SPARQLRepository class, instead of extending it. 

The main thing is to have these properties used when configuring the HttpClient used internally. You'll want these properties used to configure a `CredentialsProvider`, which should then be tacked onto the HttpClient used. An example of how this is done for basic auth is in [SPARQLProtocolSession](https://github.com/eclipse/rdf4j/blob/1f6d47982757eb192c0ac3e736ecbfd21ad23eca/core/http/client/src/main/java/org/eclipse/rdf4j/http/client/SPARQLProtocolSession.java#L313).

Edit: you could possibly also manually set your authorization header I guess. In that case you wouldn't need to use a CredentialsProvider but could instead just use the `setAdditionalHttpHeaders` method.
